File: sdks/python/glyph_protocol/pin_store.py
"""PinStore — persistent record of approved tool cards."""
from __future__ import annotations

import json
import os
import tempfile
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional


@dataclass
class Pin:
    tool_name: str
    approved_at: str
    card: dict
    revoked_at: Optional[str] = None
    revoke_reason: Optional[str] = None

    def to_dict(self) -> dict:
        d = {
            "toolName": self.tool_name,
            "approvedAt": self.approved_at,
            "card": self.card,
        }
        if self.revoked_at:
            d["revokedAt"] = self.revoked_at
        if self.revoke_reason:
            d["revokeReason"] = self.revoke_reason
        return d

    @classmethod
    def from_dict(cls, d: dict) -> "Pin":
        return cls(
            tool_name=d["toolName"],
            approved_at=d["approvedAt"],
            card=d["card"],
            revoked_at=d.get("revokedAt"),
            revoke_reason=d.get("revokeReason"),
        )


class MemoryPinStore:
    """In-memory pin store. Convenient for tests; lost on process restart."""

    def __init__(self) -> None:
        self._pins: dict[str, Pin] = {}

    def get(self, tool_name: str) -> Optional[Pin]:
        return self._pins.get(tool_name)

    def put(self, pin: Pin) -> None:
        self._pins[pin.tool_name] = pin

    def all(self) -> list[Pin]:
        return list(self._pins.values())


class FilePinStore:
    """JSON-on-disk pin store with atomic writes (write-to-temp + rename)."""

    def __init__(self, path: str) -> None:
        self.path = path

    def _load(self) -> dict[str, Pin]:
        if not os.path.exists(self.path):
            return {}
        with open(self.path, "r", encoding="utf-8") as f:
            data = json.load(f)
        return {p["toolName"]: Pin.from_dict(p) for p in data.get("pins", [])}

    def get(self, tool_name: str) -> Optional[Pin]:
        return self._load().get(tool_name)

    def all(self) -> list[Pin]:
        return list(self._load().values())

    def put(self, pin: Pin) -> None:
        pins = self._load()
        pins[pin.tool_name] = pin
        payload = {"version": 1, "pins": [p.to_dict() for p in pins.values()]}
        dirname = os.path.dirname(self.path) or "."
        fd, tmp = tempfile.mkstemp(dir=dirname, prefix=".pins-", suffix=".json")
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                json.dump(payload, f, indent=2)
                f.write("\n")
            os.replace(tmp, self.path)
        except Exception:
            try:
                os.remove(tmp)
            except OSError:
                pass
            raise

    @staticmethod
    def now() -> str:
        return datetime.now(timezone.utc).isoformat()

File: sdks/go/glyphprotocol/pinstore.go
package glyphprotocol

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sync"
)

// Pin mirrors @glyphp/types.Pin.
type Pin struct {
	ToolName     string         `json:"toolName"`
	ApprovedAt   string         `json:"approvedAt"`
	Card         map[string]any `json:"card"`
	RevokedAt    string         `json:"revokedAt,omitempty"`
	RevokeReason string         `json:"revokeReason,omitempty"`
}

// PinStore is the contract every consumer pin backend implements.
type PinStore interface {
	Get(toolName string) (*Pin, error)
	Put(pin Pin) error
	All() ([]Pin, error)
}

// MemoryPinStore is an in-process store, suitable for tests.
type MemoryPinStore struct {
	mu   sync.RWMutex
	pins map[string]Pin
}

func NewMemoryPinStore() *MemoryPinStore {
	return &MemoryPinStore{pins: map[string]Pin{}}
}

func (m *MemoryPinStore) Get(toolName string) (*Pin, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	if p, ok := m.pins[toolName]; ok {
		return &p, nil
	}
	return nil, nil
}

func (m *MemoryPinStore) Put(pin Pin) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.pins[pin.ToolName] = pin
	return nil
}

func (m *MemoryPinStore) All() ([]Pin, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make([]Pin, 0, len(m.pins))
	for _, p := range m.pins {
		out = append(out, p)
	}
	return out, nil
}

// FilePinStore persists pins to a JSON file with atomic writes.
type FilePinStore struct {
	Path string
	mu   sync.Mutex
}

type filePayload struct {
	Version int   `json:"version"`
	Pins    []Pin `json:"pins"`
}

func (f *FilePinStore) load() (map[string]Pin, error) {
	out := map[string]Pin{}
	raw, err := os.ReadFile(f.Path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return out, nil
		}
		return nil, err
	}
	var payload filePayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, err
	}
	for _, p := range payload.Pins {
		out[p.ToolName] = p
	}
	return out, nil
}

func (f *FilePinStore) Get(toolName string) (*Pin, error) {
	pins, err := f.load()
	if err != nil {
		return nil, err
	}
	if p, ok := pins[toolName]; ok {
		return &p, nil
	}
	return nil, nil
}

func (f *FilePinStore) All() ([]Pin, error) {
	pins, err := f.load()
	if err != nil {
		return nil, err
	}
	out := make([]Pin, 0, len(pins))
	for _, p := range pins {
		out = append(out, p)
	}
	return out, nil
}

func (f *FilePinStore) Put(pin Pin) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	pins, err := f.load()
	if err != nil {
		return err
	}
	pins[pin.ToolName] = pin
	payload := filePayload{Version: 1, Pins: make([]Pin, 0, len(pins))}
	for _, p := range pins {
		payload.Pins = append(payload.Pins, p)
	}
	body, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	body = append(body, '\n')
	dir := filepath.Dir(f.Path)
	if dir == "" {
		dir = "."
	}
	tmp, err := os.CreateTemp(dir, ".pins-*.json")
	if err != nil {
		return err
	}
	defer os.Remove(tmp.Name())
	if _, err := tmp.Write(body); err != nil {
		tmp.Close()
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	return os.Rename(tmp.Name(), f.Path)
}
